File: test/ModularPipelines.UnitTests/Attributes/CliToolAttributeTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.Options;
using System.Reflection;

namespace ModularPipelines.UnitTests.Attributes;

public class CliToolAttributeTests
{
    [Test]
    public async Task CliToolAttribute_StoresToolName()
    {
        var attribute = new CliToolAttribute("git");

        await Assert.That(attribute.Tool).IsEqualTo("git");
    }

    [Test]
    public async Task CliToolAttribute_CanBeAppliedToClass()
    {
        var attribute = typeof(TestGitOptions).GetCustomAttribute<CliToolAttribute>();

        await Assert.That(attribute).IsNotNull();
        await Assert.That(attribute!.Tool).IsEqualTo("git");
    }

    [Test]
    public async Task CliToolAttribute_ThrowsOnNullOrWhitespace()
    {
        await Assert.That(() => new CliToolAttribute(null!)).Throws<ArgumentException>();
        await Assert.That(() => new CliToolAttribute("")).Throws<ArgumentException>();
        await Assert.That(() => new CliToolAttribute("   ")).Throws<ArgumentException>();
    }

    [Test]
    public async Task CliToolAttribute_IsInheritedByDerivedClasses()
    {
        var attribute = typeof(TestGitCommitOptions).GetCustomAttribute<CliToolAttribute>(inherit: true);

        await Assert.That(attribute).IsNotNull();
        await Assert.That(attribute!.Tool).IsEqualTo("git");
    }

    [CliTool("git")]
    private abstract record TestGitOptions : CommandLineToolOptions;

    private record TestGitCommitOptions : TestGitOptions;

    [CliTool("test")]
    private record TestOptionsWithAttribute : CommandLineToolOptions;
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/CliToolAttributeTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open System.Reflection
open ModularPipelines.Attributes
open ModularPipelines.Options
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

[<AbstractClass>]
[<CliTool("git")>]
type private TestGitOptions() =
    inherit CommandLineToolOptions()

[<AbstractClass>]
type private TestGitCommitOptions() =
    inherit TestGitOptions()

[<AbstractClass>]
[<CliTool("test")>]
type private TestOptionsWithAttribute() =
    inherit CommandLineToolOptions()

type CliToolAttributeTests() =
    [<Test>]
    member _.CliToolAttribute_StoresToolName() = async {
        let attribute = CliToolAttribute("git")
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(attribute.Tool), "git"))
    }

    [<Test>]
    member _.CliToolAttribute_CanBeAppliedToClass() = async {
        let attribute = typeof<TestGitOptions>.GetCustomAttribute<CliToolAttribute>()
        do! check(Assert.That(attribute).IsNotNull())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(attribute.Tool), "git"))
    }

    [<Test>]
    member _.CliToolAttribute_ThrowsOnNullOrWhitespace() = async {
        let mutable threw1 = false
        try
            CliToolAttribute(null) |> ignore
        with :? System.ArgumentException ->
            threw1 <- true

        let mutable threw2 = false
        try
            CliToolAttribute("") |> ignore
        with :? System.ArgumentException ->
            threw2 <- true

        let mutable threw3 = false
        try
            CliToolAttribute("   ") |> ignore
        with :? System.ArgumentException ->
            threw3 <- true

        do! check(Assert.That(threw1).IsTrue())
        do! check(Assert.That(threw2).IsTrue())
        do! check(Assert.That(threw3).IsTrue())
    }

    [<Test>]
    member _.CliToolAttribute_IsInheritedByDerivedClasses() = async {
        let attribute = typeof<TestGitCommitOptions>.GetCustomAttribute<CliToolAttribute>(true)
        do! check(Assert.That(attribute).IsNotNull())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(attribute.Tool), "git"))
    }
